pkg/storage/s3/s3.go

```diff
@@ -0,0 +1,79 @@
+package s3
+
+import (
+	"fmt"
+	"net/url"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3/s3iface"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager/s3manageriface"
+	"github.com/gomods/athens/pkg/config"
+	"github.com/gomods/athens/pkg/errors"
+)
+
+// Storage implements (./pkg/storage).Backend and
+// also provides a function to fetch the location of a module
+// Storage uses amazon aws go SDK which expects these env variables
+// - AWS_REGION 			- region for this storage, e.g 'us-west-2'
+// - AWS_ACCESS_KEY_ID		-
+// - AWS_SECRET_ACCESS_KEY 	-
+// - AWS_SESSION_TOKEN		- [optional]
+// For information how to get your keyId and access key turn to official aws docs: https://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/setting-up.html
+type Storage struct {
+	bucket   string
+	baseURI  *url.URL
+	uploader s3manageriface.UploaderAPI
+	s3API    s3iface.S3API
+	s3Conf   *config.S3Config
+	cdnConf  *config.CDNConfig
+}
+
+// New creates a new AWS S3 CDN saver
+func New(s3Conf *config.S3Config, cdnConf *config.CDNConfig, options ...func(*aws.Config)) (*Storage, error) {
```

Comment:
what's the purpose of the options here?

Reply:
Used mostly for tests to configure additional fields, like endpoint and disabled SSL. They should be on default when working with normal S3

Reply:
if it's only for test, can test prepare config it would like to have in Arrange step and clean after itself afterward if needed?
i'm not a fan of tweaking implementation for sake of tests

Reply:
Yeah but then I would need to rollback this uneccessary fields that where before in config.S3Config. I don't see the way how to set that options before creating session to S3, I could create antoher `New` method eg. `NewWithConfig/NewWithOptions` and then pass this additional settings but that will create method solely for tests.